scripts/cot/botania/flowers/genereting/campanimia.zs

```diff
@@ -0,0 +1,168 @@
+import mods.randomtweaker.cote.Update;
+import crafttweaker.data.IData;
+import mods.zenutils.DataUpdateOperation.APPEND;
+import mods.zenutils.DataUpdateOperation.MERGE;
+import mods.zenutils.DataUpdateOperation.REMOVE;
+import mods.zenutils.DataUpdateOperation.OVERWRITE;
+import crafttweaker.world.IWorld;
+import crafttweaker.block.IBlock;
+import crafttweaker.world.IBlockPos;
+import crafttweaker.util.Position3f;
+import crafttweaker.player.IPlayer;
+import crafttweaker.entity.IEntity;
+import crafttweaker.util.Math;
+
+var campanimia as ISubTileEntityGenerating = VanillaFactory.createSubTileGenerating("campanimia", 65535);
+campanimia.maxMana = 300000; 
+campanimia.passiveFlower = false;
+campanimia.range = 1;
+campanimia.onUpdate = function(subtile, world, pos) { 
+    if(world.isRemote()) return;
+    if(world.time%20 != 7) return; 
+    val cruciblesPos = getCruciblesPos(world, pos);
+    if(cruciblesPos.length == 0) return;
+    val crucibles = getCrucibles(world, cruciblesPos);
+    val aspects = getAspectsList(crucibles);
+    if(aspects.length == 0) return;
+    val generatedMana = pow(10,pow(aspects.length,0.475) - 1) as int;
+    if(subtile.getMaxMana() - subtile.getMana() < generatedMana) return;
+    drinkEssentia(world, cruciblesPos);
+    subtile.addMana(generatedMana);
+    playSound("thaumcraft:spill", pos, world);
+    return;
+};
+campanimia.register();
+
+function getCruciblesPos(world as IWorld, posFlower as IBlockPos) as IBlockPos[] {
+    var cruciblesPos = [] as IBlockPos[];
+    val poss = [
+        toBlockPos(posFlower.x + 1, posFlower.y, posFlower.z),
+        toBlockPos(posFlower.x - 1, posFlower.y, posFlower.z),
+        toBlockPos(posFlower.x, posFlower.y, posFlower.z + 1),
+        toBlockPos(posFlower.x, posFlower.y, posFlower.z - 1),
+    ] as IBlockPos[];
+
+    for pos in poss {
+        val block = world.getBlock(pos);
+        if(isNull(block)
+        || isNull(block.data)
+        || isNull(block.data.id)
+        || block.data.id!="thaumcraft:tilecrucible"
+        || block.data.Amount<100
+        || block.data.Heat!=200
+        || block.data.Aspects.length==0) continue;
+        cruciblesPos += pos;
+    }
+
+    return cruciblesPos;
+}
+
+function toBlockPos(x as int, y  as int, z  as int) as IBlockPos{
+    return crafttweaker.util.Position3f.create(x, y, z) as IBlockPos;
+}
+
+function getCrucibles(world as IWorld, cruciblesPosList as IBlockPos[]) as IBlock[]{
+    var crucibles = [] as IBlock[];
+    for cruciblePos in cruciblesPosList{
+        crucibles += world.getBlock(cruciblePos);
+    }
+    return crucibles;
+}
+
+function getAspectsList(crucibles as IBlock[]) as IData { 
+    var aspects as IData = [];
+    for crucible in crucibles {
+        for i in 0 .. crucible.data.Aspects.length{
+            aspects += [crucible.data.Aspects[i].key];
+        }
+    }
+    return aspects;
+}
+
+function drinkEssentia(world as IWorld, cruciblesPosList as IBlockPos[]) as void {
```

Comment:
Please make this tests:

1. Place and configure 20 flowers of each added type
2. Automate them to work continiously
3. Create Spark Profile
4. Send profile to comment here

I want to see if all flower functions are performant enough.

Reply:
https://spark.lucko.me/Yy75xxy6f6

I'm not sure if this is what was needed

Reply:
Yes, this is what i need.
Profile shows that Antirrhift is causing small lag. Its notable, but not terrible. If you can optimize it - would be good.
As you can see on the call stack, block nbt get/set is what causing most of CPU strain. Possible solution would be delay updates OR some high-skill native manipulations, like directly changing TileEntity class properties, omitting working with NBT.

![image](https://github.com/user-attachments/assets/65e1bb59-65f2-444b-b9f0-65ff3c3294a5)